### PR TITLE
curvefs/client : Performance optimization

### LINKED ISSUE #704: reduce lock conflicts
**Is your feature request related to a problem? (你需要的功能是否与某个问题有关?)**
after every datacache flush， there will update inode.

**Describe the solution you'd like (描述你期望的解决方法)**
update inode is heavy operation, it will apply for the lock.
we can update inode just in every chunk flush

**Describe alternatives you've considered (描述你想到的折衷方案)**

**Additional context/screenshots (更多上下文/截图)**

### PR DESCRIPTION
**Pull request by wuhongsong**
Issue Number: close #704

### CODE REVIEW

**Comment by wuhongsong** [comment]
recheck

**Comment by wuhongsong** [comment]
recheck

File: curvefs/src/client/s3/client_s3_cache_manager.cpp [hunk 1]
```diff
@@ -1283,9 +1294,74 @@ CURVEFS_ERROR ChunkCacheManager::Flush(uint64_t inodeId, bool force) {
...
+        } else {
+            S3ChunkInfoList s3ChunkInfoList;
+            UpdateInodeChunkInfo(&s3ChunkInfoList,
+                chunkId, offset, writeOffset);
+            s3ChunkInfoMap->insert({chunkIndex, s3ChunkInfoList});
+        }
+        s3chunkInfoListIter = s3ChunkInfoMap->find(chunkIndex);
+        exist = true;
+    }
+    inodeWrapper->SwapInode(&inode);
```
Comment (xu-chaojie) [inline curvefs/src/client/s3/client_s3_cache_manager.cpp:1349]: rebase loose change

File: curvefs/src/client/s3/client_s3_cache_manager.cpp [hunk 2]
```diff
@@ -1252,19 +1252,27 @@ void ChunkCacheManager::ReleaseWriteDataCache(uint64_t key,
...
     {
         WriteLockGuard writeLockGuard(rwLockWrite_);
         tmp = dataWCacheMap_;
     }
+    std::vector<uint64_t> chunkIdList;
+    std::vector<uint64_t> offsetList;
+    std::vector<uint64_t> writeOffsetList;
+    uint64_t chunkId;
+    uint64_t offset;
+    uint32_t writeOffset;
```
Comment (xu-chaojie) [inline curvefs/src/client/s3/client_s3_cache_manager.cpp:1268]: why uint32_t here?

File: curvefs/src/client/s3/client_s3_cache_manager.cpp [hunk 3]
```diff
@@ -1283,9 +1294,74 @@ CURVEFS_ERROR ChunkCacheManager::Flush(uint64_t inodeId, bool force) {
...
+            inodeId, inodeWrapper);
+    if (ret != CURVEFS_ERROR::OK) {
+        LOG(WARNING) << "get inode fail, ret:" << ret;
+        return ret;
+    }
+    uint64_t chunkId;
+    uint64_t offset;
+    uint32_t writeOffset;
+    ::curve::common::UniqueLock lgGuard = inodeWrapper->GetUniqueLock();
+    Inode inode = inodeWrapper->GetInodeUnlocked();
```
Comment (xu-chaojie) [inline curvefs/src/client/s3/client_s3_cache_manager.cpp:1323]: rebase lost change here

File: curvefs/src/client/s3/client_s3_cache_manager.cpp [hunk 4]
```diff
@@ -1283,9 +1294,74 @@ CURVEFS_ERROR ChunkCacheManager::Flush(uint64_t inodeId, bool force) {
...
+            S3ChunkInfoList &s3ChunkInfoList = s3chunkInfoListIter->second;
+            UpdateInodeChunkInfo(&s3ChunkInfoList, chunkId,
+              offset, writeOffset);
+        } else {
+            S3ChunkInfoList s3ChunkInfoList;
+            UpdateInodeChunkInfo(&s3ChunkInfoList,
+                chunkId, offset, writeOffset);
+            s3ChunkInfoMap->insert({chunkIndex, s3ChunkInfoList});
+        }
+        s3chunkInfoListIter = s3ChunkInfoMap->find(chunkIndex);
```
Comment (xu-chaojie) [inline curvefs/src/client/s3/client_s3_cache_manager.cpp:1346]: should we move those two line before }

File: curvefs/src/client/s3/client_s3_cache_manager.cpp [hunk 5]
```diff
@@ -1529,38 +1606,7 @@ CURVEFS_ERROR DataCache::Flush(uint64_t inodeId, bool force) {
...
-            } else {
-                S3ChunkInfoList &s3ChunkInfoList = s3chunkInfoListIter->second;
-                UpdateInodeChunkInfo(&s3ChunkInfoList, chunkId, offset,
-                                     writeOffset);
-            }
-            s3ClientAdaptor_->GetInodeCacheManager()->ShipToFlush(inodeWrapper);
-        }
-
-        VLOG(9) << "data flush end, inodeId: " << inodeId;
+        delete data;
```
Comment (ilixiaocui) [inline curvefs/src/client/s3/client_s3_cache_manager.cpp:1609]: if (data != nullptr)
delete []data

Comment (wuhongsong) [reply curvefs/src/client/s3/client_s3_cache_manager.cpp:1609]: fix

File: curvefs/src/client/s3/client_s3_cache_manager.cpp (hunk 1 above)
Comment (wuhongsong) [reply curvefs/src/client/s3/client_s3_cache_manager.cpp:1319]: fix

File: curvefs/src/client/s3/client_s3_cache_manager.cpp [hunk 6]
```diff
@@ -1252,19 +1252,27 @@ void ChunkCacheManager::ReleaseWriteDataCache(uint64_t key,
...
+    VLOG(9) << "chunk flush start, inodeId: " << inodeId;
     std::map<uint64_t, DataCachePtr> tmp;
     curve::common::LockGuard lg(flushMtx_);
     CURVEFS_ERROR ret;
     {
         WriteLockGuard writeLockGuard(rwLockWrite_);
         tmp = dataWCacheMap_;
     }
+    std::vector<uint64_t> chunkIdList;
+    std::vector<uint64_t> offsetList;
```
[the remark this answers is not in the document]
Comment (wuhongsong) [reply curvefs/src/client/s3/client_s3_cache_manager.cpp:1220]: offsetList is store offset, and writeOffsetList is store writeoffset. offset is the offset in file, and writeoffset is the offset after one write

File: curvefs/src/client/s3/client_s3_cache_manager.cpp [hunk 7]
```diff
@@ -1252,19 +1252,27 @@ void ChunkCacheManager::ReleaseWriteDataCache(uint64_t key,
...
+    std::vector<uint64_t> writeOffsetList;
+    uint64_t chunkId;
```
[the remark this answers is not in the document]
Comment (wuhongsong) [reply curvefs/src/client/s3/client_s3_cache_manager.cpp:1222]: fix

File: curvefs/src/client/s3/client_s3_cache_manager.cpp (hunk 2 above)
Comment (wuhongsong) [reply curvefs/src/client/s3/client_s3_cache_manager.cpp:1224]: fix

File: curvefs/src/client/s3/client_s3_cache_manager.cpp [hunk 8]
```diff
@@ -1283,9 +1294,74 @@ CURVEFS_ERROR ChunkCacheManager::Flush(uint64_t inodeId, bool force) {
...
+    if (ret != CURVEFS_ERROR::OK) {
+        LOG(ERROR) << "update inode fail, ret: " << ret;
+        return ret;
+    }
+    VLOG(9) << "chunk flush end, inodeId: " << inodeId;
+    return CURVEFS_ERROR::OK;
+}
+
+CURVEFS_ERROR ChunkCacheManager::UpdateInode(uint64_t inodeId,
+  std::vector<uint64_t> chunkIdList, std::vector<uint64_t> offsetList,
```
[the remark this answers is not in the document]
Comment (wuhongsong) [reply curvefs/src/client/s3/client_s3_cache_manager.cpp:1277]: fix

File: curvefs/src/client/s3/client_s3_cache_manager.cpp [hunk 9]
```diff
@@ -1283,9 +1294,74 @@ CURVEFS_ERROR ChunkCacheManager::Flush(uint64_t inodeId, bool force) {
...
+    uint64_t chunkIndex = GetIndex();
+    std::shared_ptr<InodeWrapper> inodeWrapper;
+    CURVEFS_ERROR ret =
+        s3ClientAdaptor_->GetInodeCacheManager()->GetInode(
+            inodeId, inodeWrapper);
+    if (ret != CURVEFS_ERROR::OK) {
+        LOG(WARNING) << "get inode fail, ret:" << ret;
+        return ret;
+    }
+    uint64_t chunkId;
```
[the remark this answers is not in the document]
Comment (wuhongsong) [reply curvefs/src/client/s3/client_s3_cache_manager.cpp:1289]: fix

File: curvefs/src/client/s3/client_s3_cache_manager.cpp (hunk 4 above)
Comment (wuhongsong) [reply curvefs/src/client/s3/client_s3_cache_manager.cpp:1316]: fix

File: curvefs/src/client/s3/client_s3_cache_manager.cpp [hunk 10]
```diff
@@ -1410,21 +1486,22 @@ void DataCache::Release() {
     return;
 }
 
-CURVEFS_ERROR DataCache::Flush(uint64_t inodeId, bool force) {
+CURVEFS_ERROR DataCache::Flush(uint64_t inodeId, bool force,
+  uint64_t *chunkId, uint64_t* offset, uint32_t* writeOffset) {
```
[the remark this answers is not in the document]
Comment (wuhongsong) [reply curvefs/src/client/s3/client_s3_cache_manager.cpp]: fix

**Pull request closed without merging**